File: hdf5_2/C/solution/hdf5_ex2.c
/* ###################################################################
   #                       HDF5 hands on                             #
   #  Created : April 2011                                           #
   #                                                                 #
   #  Author:                                                        #
   #     Matthieu Haefele                                            #
   #     matthieu.haefele@ipp.mpg.de                                 #
   #     High Level Support Team (HLST)                              #
   #     Max-Planck Institut fuer Plasmaphysik                       #
   #                                                                 #
   ################################################################### */

#include <hdf5.h>
#define NX     10
#define NY     6
#define NZ     3

void init(int* data, int* data3d)
{
  int i,j,k;
  for(j=0; j < NY; j++)
    for(i=0; i < NX; i++)
      data[i + j*NX] = i + j*NX;

  for(k=0; k < NZ; k++)	  
    for(j=0; j < NY; j++)
      for(i=0; i < NX; i++)
        data3d[i + j*NX + k*NX*NY] = i + j*NX + k*NX*NY;
}

int main (void)
{
  hid_t       file, dataset, dataspace;
  hsize_t     dimsf[3];
  hsize_t array_size[3], array_subsize[3], array_start[3];
  herr_t      status;
  int         data[NX*NY];
  int         data3d[NX*NY*NZ];
  
  //Data initialization
  init(data,data3d);
  
  //HDF5 file creation
  file = H5Fcreate("example.h5", H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
  dimsf[0] = NY;
  dimsf[1] = NX;
  
  //Dataspace creation
  dataspace = H5Screate_simple(2, dimsf, NULL); 
  
  //Dataset creation 
  dataset = H5Dcreate(file, "IntArray", H5T_NATIVE_INT, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

  //Actual data IO
  status = H5Dwrite(dataset, H5T_NATIVE_INT, H5S_ALL, H5S_ALL,H5P_DEFAULT, data);
  
  //Closing dataset
  H5Dclose(dataset);

  //Dataset creation for the second 2D dataset by reusing the dataspace
  dataset = H5Dcreate(file, "IntArray2", H5T_NATIVE_INT, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

  //Closing dataspace
  H5Sclose(dataspace);

  //Dataspace creation for memory selection
  dimsf[0] = NZ;
  dimsf[1] = NY;
  dimsf[2] = NX;
  array_subsize[0] = 1;
  array_subsize[1] = NY;
  array_subsize[2] = NX;
  array_start[0] = 2; 
  array_start[1] = 0; 
  array_start[2] = 0; 
  dataspace = H5Screate_simple(3, dimsf, NULL); 
  status = H5Sselect_hyperslab (dataspace, H5S_SELECT_SET, array_start, NULL, array_subsize, NULL);

  //Actual data IO
  status = H5Dwrite(dataset, H5T_NATIVE_INT, dataspace, H5S_ALL,H5P_DEFAULT, data3d);
  
  //Closing all opened HDF5 objects
  H5Sclose(dataspace);
  H5Dclose(dataset);
  H5Fclose(file);
  

  return 0;
}     

File: hdf5_2/F90/solution/hdf5_ex2.f90

! ###################################################################
! #                       HDF5 hands on                             #
! #  Created : April 2011                                           # 
! #                                                                 #
! #  Author:                                                        #
! #     Matthieu Haefele                                            #
! #     matthieu.haefele@ipp.mpg.de                                 #
! #     High Level Support Team (HLST)                              #
! #     Max-Planck Institut fuer Plasmaphysik                       #
! #                                                                 #
! ###################################################################

PROGRAM HDF5_EX
     USE HDF5

     IMPLICIT NONE

     CHARACTER(LEN=10), PARAMETER :: filename = "example.h5" ! File name
     CHARACTER(LEN=8), PARAMETER :: dsetname = "IntArray"     ! Dataset name
     CHARACTER(LEN=10), PARAMETER :: dsetname2= "IntArray2"     ! Dataset name
     INTEGER, PARAMETER     ::   NX = 10                        
     INTEGER, PARAMETER     ::   NY = 6                        
     INTEGER, PARAMETER     ::   NZ = 3                        

     INTEGER(HID_T) :: file, dataset, dataspace
     INTEGER(HSIZE_T), DIMENSION(3) :: dims, array_subsize, array_start
     INTEGER     ::   status
     INTEGER, DIMENSION(NX,NY) :: data
     INTEGER, DIMENSION(NX,NY,NZ) :: data3d
     INTEGER     :: i,j,k
     
     DO j=1,NY
       DO i=1,NX
        data(i,j) = i-1 + (j-1)*NX;
       ENDDO
     ENDDO
        
     DO k=1,NZ
       DO j=1,NY
         DO i=1,NX
           data3d(i,j,k) = i-1 + (j-1)*NX + (k-1)*NX*NY
         ENDDO
       ENDDO
     ENDDO
        
     !
     ! Initialize FORTRAN predefined datatypes.
     !
     CALL h5open_f(status)

     !
     ! Create a new file using default properties.
     ! 
     CALL h5fcreate_f(filename, H5F_ACC_TRUNC_F, file, status)
     
     
     ! 
     ! Create the dataspace.
     !
     dims(1) = NX
     dims(2) = NY
     CALL h5screate_simple_f(2, dims, dataspace, status)
     
     
     !
     ! Create the dataset default properties.
     !
     CALL h5dcreate_f(file, dsetname, H5T_NATIVE_INTEGER, dataspace, &
                      dataset, status)
      
     !
     ! Write the data actually
     !
     CALL h5dwrite_f(dataset, H5T_NATIVE_INTEGER, data, dims, status)


     !   
     ! End access to the dataset and release resources used by it.
     ! 
     CALL h5dclose_f(dataset, status)

     !
     ! Create the dataset for the second 2d array by reusing the dataspace
     !
     CALL h5dcreate_f(file, dsetname2, H5T_NATIVE_INTEGER, dataspace, &
                      dataset, status)
      
     !
     ! Terminate access to the data space.
     !
     CALL h5sclose_f(dataspace, status)

     ! Create the 3D dataspace to make the selection in memory.
     !
     dims(1) = NX
     dims(2) = NY
     dims(3) = NZ
     CALL h5screate_simple_f(3, dims, dataspace, status)
     array_subsize(1) = NX
     array_subsize(2) = NY
     array_subsize(3) = 1
     array_start(1) = 0 ! C convention 
     array_start(2) = 0 ! C convention 
     array_start(3) = 2 ! C convention 
     CALL h5sselect_hyperslab_f (dataspace, H5S_SELECT_SET_F, array_start, array_subsize, status)

    
     
     !
     ! Write the data actually
     !
     CALL h5dwrite_f(dataset, H5T_NATIVE_INTEGER, data3d, dims, status, mem_space_id=dataspace)


     !   
     ! End access to the dataset and release resources used by it.
     ! 
     CALL h5dclose_f(dataset, status)

     !
     ! Terminate access to the data space.
     !
     CALL h5sclose_f(dataspace, status)

     ! 
     ! Close the file.
     !
     CALL h5fclose_f(file, status)

     !
     ! Close FORTRAN predefined datatypes.
     !
     CALL h5close_f(status)

     END PROGRAM HDF5_EX
     
